pricegraph/src/orderbook.rs

```diff
@@ -140,6 +141,69 @@ impl Orderbook {
         debug_assert!(!self.is_overlapping());
     }
 
+    /// Reduces all overlapping orders for a given market and returns the
+    /// computed overlapping transitive orderbook for that market.
+    pub fn reduce_overlapping_transitive_orderbook(
+        &mut self,
+        base: TokenId,
+        quote: TokenId,
+    ) -> TransitiveOrderbook {
+        let (base, quote) = (node_index(base), node_index(quote));
+
+        let mut overlap = TransitiveOrderbook::default();
+        while let Err(NegativeCycle(predecessors, node)) =
+            bellman_ford::search(&self.projection, quote)
+        {
+            let path = {
+                let mut cycle = path::find_cycle(&predecessors, node, Some(quote))
+                    .expect("negative cycle not found after being detected");
+                cycle.push(cycle[0]);
+                cycle
+            };
+
+            if path[0] == quote {
+                if let Some(base_index) = path.iter().position(|node| *node == base) {
+                    let (ask, bid) = (&path[0..base_index + 1], &path[base_index..]);
+
+                    overlap.asks.push(
+                        self.find_transitive_order(ask)
+                            .expect("ask transitive path not found after being detected"),
+                    );
+                    overlap.bids.push(
+                        self.find_transitive_order(bid)
+                            .expect("bid transitive path not found after being detected"),
+                    );
+                }
+            }
+
+            self.fill_path(&path).unwrap_or_else(|| {
+                panic!(
+                    "failed to fill path along detected negative cycle {}",
+                    format_path(&path),
+                )
+            });
+        }
+
+        overlap
+    }
+
+    /// Computes the equivalent transitive order along a path.
+    ///
+    /// Returns `None` if the path does not exist.
+    fn find_transitive_order(&self, path: &[NodeIndex]) -> Option<TransitiveOrder> {
+        let (capacity, price) = self.find_path_capacity_and_price(path)?;
+
+        // NOTE: We now have the capacity and price for this transitive order
+        // which needs to be converted to a buy and sell amount. We have:
+        // - `price = FEE_FACTOR * buy_amount / sell_amount`
+        // - `capacity = sell_amount * price`
+        // Solving for `buy_amount` and `sell_amount`, we get:
+        let buy = capacity / FEE_FACTOR;
```

Comment:
Just to double check, the fees for each sub-order of the path is already considered in the original orders, right?

Reply:
They are considered in the prices since they are calculated with `price = FEE_FACTOR * buy_amount / sell_amount`. So that means that if the transitive order "hops" over 3 orders (`A -> B -> C -> D`, each arrow being an order), for example, it would introduce fees of `FEE_FACTOR ** 3` (one for each hop). The last one we deduct, since orders implicitly include fees - so in our example, we would end up with `FEE_FACTOR ** 2`, which from what I gather is what we want.